Feature/organize and cache

theherk opened the pull request:
This pull request represents a significant modification to how the code is organized. It is also a shift in how the cache is handled.
### Organization

The biggest change is how the code is organized. The methods are now grouped by category, similar to how Commvault groups them in the api. See [here](http://documentation.commvault.com/commvault/v10/article?p=features/rest_api/rest_api_libraries.htm)

These are all initialized by the commvault module, but they can be initialized alone if you only need a subset of the methods.

The base_session module handles authentication and requests, as well as the implementation of the cache.
### Compatibility

Although effort was made to make it almost 100% backward compatible, one breaking change was made. `get_job_details` previously required `client_id`. That variable was unused, and therefore, removed. This change the api and will require a major point increase.

The wrapper was written for v10 but is now being update to support v11. However, tests are still required.
### Cache

The caching is now done in LRU cache but base on TTL. You can pass in which methods will have caching enable and a value, in seconds, for the TTL. However, great defaults are provided.

When passing in a list of methods to be cached, don't worry about which module they are in, just name the methods. They will be found.

theherk committed:
Move errors to separate module

theherk committed:
Move api methods to organized modules

theherk committed:
Init api modules

theherk committed:
Add shim for backward compatibility

theherk committed:
Move remaining methods to base_session and handle caching

defond0 commented on pinkopy/base_session.py:
```diff
@@ -0,0 +1,183 @@
+
+
+class BaseSession(object):
+    """BaseSession
```
love the BaseSession

defond0 commented on pinkopy/base_session.py:
```diff
@@ -0,0 +1,183 @@
+        try:
+            method = getattr(self, method_name)
+            setattr(self, method_name, ttl_cache(ttl=self.cache_ttl)(method))
+            return True
```
this method is pure gold

theherk committed:
Remove unused imports

theherk committed:
Modify responses to support Commvault v10 and v11

This included making one requests via xml. They seem to have broken some
of the json requests.

theherk committed:
Move context manager methods to base session

theherk committed:
Share authtoken among sessions

This way only one login session happens. The logout is shadowed on
commvault session, to logout each of the subsessions.

theherk committed:
Ensure that no method is wrapped in cache more than once

theherk committed:
Add cache_methods to commvault

This way we can honestly reply about the methods that have been cached.
This is set double comprehension cast to a list.

theherk committed:
Initialize the super class so that commvault has correct properties

This way commvault.use_cache and friends have the proper data to report
to the user.
